README.rst
Async-Slumber |build-status| |coverage-status| |docs|
===============================================

Slumber is a Python library that provides a convenient yet powerful
object-oriented interface to ReSTful APIs. It acts as a wrapper around the
excellent aiohttp_ library and abstracts away the handling of URLs, serialization,
and request processing.

.. _aiohttp: http://aiohttp.readthedocs.io/en/stable/

QuickStart
==========

1. Install Slumber::

    $ pip install slumber

2. Install Optional Requirement::

    pip install pyyaml

3. Use Slumber!

Requirements
============

Slumber requires the following modules.

* Python 3.5+
* aiohttp
* pyyaml (If you are using the optional YAML serialization)

.. |build-status| image:: https://travis-ci.org/samgiles/slumber.svg?branch=master
   :target: https://travis-ci.org/samgiles/slumber
   :alt: Build status
.. |coverage-status| image:: https://img.shields.io/coveralls/samgiles/slumber.svg
   :target: https://coveralls.io/r/samgiles/slumber
   :alt: Test coverage percentage
.. |docs| image:: https://readthedocs.org/projects/slumber/badge/?version=latest
   :target: http://slumber.readthedocs.org/
   :alt: Documentation


CHANGELOG.rst
.. :changelog:

Changelog
=========

2017.7
------
* ``aiohttp`` in place of requests
* Support to Python 3.5 only for now.


development version
-------------------

0.7.1
-----

* Correct support for UTF encoded responses.

0.6.2
-----

* Further Python 3 compatibility fixes, many thanks to Alexander Shchepetilnikov (irqed)

0.6.1
-----

* Remove ``simplejson`` dependency.

* Support range now Python 2.6 - Python 3.4


0.6.0
-----

* All HTTP methods support multipart/form-data.  The new ``files``
  parameter is directly passed to ``requests``.  Thanks to Philip
  Neustrom and Diego Gaustein.


0.5.2
-----

* Now compatible with ``requests`` 1.0 thanks to Matias Saguir.

* Fix bug when using default serializer.  Thanks to Andy McKay


0.4.2
-----

* Support decoded unicode url fragments - Thanks @collinwat


0.4.1
-----

* Added a ``session`` kwarg to ``slumber.API`` allowing passing a ``requests`` session
  that will be used instead of the slumber created one

0.4.0
-----

* *(Backwards Incompatible)* Switched from ``httplib2`` to ``requests``
* *(Backwards Incompatible)* Removed the Meta class Magic
* *(Backwards Incompatible)* Removed the ability to subclass ``slumber.API`` to specify defaults
* *(Backwards Incompatible)* New Syntax for Specifying Authentication

0.3.1
-----

* Fix regression where pre 0.3 urls were assumed to end in slash, and 0.3.0 presumed to end in not slash.
  Urls are now assumed to end in a slash, and if you don't want this behavior you can disable it by the
  append_slash kwarg/Meta option (set to False to disable it).
* Fix regression caused by a mistyped variable name.

0.3.0
-----

* Allowed nesting resources infinitely to allow more complex api usage.
* Cleaned up the Meta class and allow subclassing ``slumber.API``
* *(Backwards Incompatible)* Cleaned up the exception names.
* *(Backwards Incompatible)* Renamed the ``slumber.API`` serialization kwarg from
  default_format to format to be more consistent
* Improved the documentation
* Added Some Tests (This could still be better)

0.2.5
-----

* Fixed https urls and the accidental force to port 80
* Fixed the assumption that all urls end in a trailing slash

0.2.4
-----

* Fixed Including of Changelog.rst

0.2.3
-----

* Updated the docs to include a section about url parameters

0.2
----

* *(Backwards Incompatible)* Move specifying a non default serializer from
  ``api.resource.get(format="yaml")`` to ``api.resource(format="yaml").get()``

* Reworked the internal ``Resource`` api to not clobber any kwargs passed to it. This
  fixes a bug where you couldn't use ``format`` or ``url`` as the name for one of
  the url parameters.

0.1.3
-----

* Fix for ``Resource.post()`` not passing kwargs to ``Resource.get()``

0.1.2
-----

* Initial public release of Slumber


tests/resource.py
# -*- coding: utf-8 -*-
import sys
import mock
import requests
import aiohttp
import slumber
import slumber.serialize
import unittest2 as unittest

from slumber import exceptions


class ResourceTestCase(unittest.TestCase):

    def setUp(self):
        self.base_resource = slumber.Resource(
                base_url="http://example/api/v1/test", format="json",
                append_slash=False, raw=False)

    def test_get_200_json(self):
        r = mock.Mock(spec=aiohttp.ClientResponse)
        r.status_code = 200
        r.headers = {"content-type": "application/json"}
        r.content = '{"result": ["a", "b", "c"]}'

        self.base_resource._store.update({
            "session": mock.Mock(spec=aiohttp.ClientSession),
            "serializer": slumber.serialize.Serializer(),
        })
        self.base_resource._store["session"].request.return_value = r

        resp = self.base_resource._request("GET")

        self.assertTrue(resp is r)
        self.assertEqual(resp.content, r.content)

        self.base_resource._store["session"].request.assert_called_once_with(
            "GET",
            "http://example/api/v1/test",
            data=None,
            files=None,
            params=None,
            headers={"accept": self.base_resource._store["serializer"].get_content_type()}
        )

        resp = self.base_resource.get()
        self.assertEqual(resp['result'], ['a', 'b', 'c'])

    def test_get_200_text(self):
        r = mock.Mock(spec=aiohttp.ClientResponse)
        r.status_code = 200
        r.headers = {"content-type": "text/plain"}
        r.content = "Mocked Content"

        self.base_resource._store.update({
            "session": mock.Mock(spec=aiohttp.ClientSession),
            "serializer": slumber.serialize.Serializer(),
        })
        self.base_resource._store["session"].request.return_value = r

        resp = self.base_resource._request("GET")

        self.assertTrue(resp is r)
        self.assertEqual(resp.content, "Mocked Content")

        self.base_resource._store["session"].request.assert_called_once_with(
            "GET",
            "http://example/api/v1/test",
            data=None,
            files=None,
            params=None,
            headers={"accept": self.base_resource._store["serializer"].get_content_type()}
        )

        resp = self.base_resource.get()
        self.assertEqual(resp, r.content)

    def test_options_200_json(self):
        r = mock.Mock(spec=aiohttp.ClientResponse)
        r.status_code = 200
        r.headers = {"content-type": "application/json"}
        r.content = '{"actions": {"POST": {"foo": {"required": false, "type": "string"}}}}'

        self.base_resource._store.update({
            "session": mock.Mock(spec=aiohttp.ClientSession),
            "serializer": slumber.serialize.Serializer(),
        })
        self.base_resource._store["session"].request.return_value = r

        resp = self.base_resource._request("OPTIONS")

        self.assertTrue(resp is r)
        self.assertEqual(resp.content, r.content)

        self.base_resource._store["session"].request.assert_called_once_with(
            "OPTIONS",
            "http://example/api/v1/test",
            data=None,
            files=None,
            params=None,
            headers={"accept": self.base_resource._store["serializer"].get_content_type()}
        )

        resp = self.base_resource.options()
        self.assertTrue('POST' in resp['actions'])
        self.assertTrue('foo' in resp['actions']['POST'])
        self.assertTrue('type' in resp['actions']['POST']['foo'])
        self.assertEqual(resp['actions']['POST']['foo']['type'], 'string')

    def test_head_200_json(self):
        r = mock.Mock(spec=aiohttp.ClientResponse)
        r.status_code = 200
        r.headers = {"content-type": "application/json"}
        r.content = ''

        self.base_resource._store.update({
            "session": mock.Mock(spec=aiohttp.ClientSession),
            "serializer": slumber.serialize.Serializer(),
        })
        self.base_resource._store["session"].request.return_value = r

        resp = self.base_resource._request("HEAD")

        self.assertTrue(resp is r)
        self.assertEqual(resp.content, r.content)

        self.base_resource._store["session"].request.assert_called_once_with(
            "HEAD",
            "http://example/api/v1/test",
            data=None,
            files=None,
            params=None,
            headers={"accept": self.base_resource._store["serializer"].get_content_type()}
        )

        resp = self.base_resource.head()
        self.assertEqual(resp, r.content)

    def test_post_201_redirect(self):
        r1 = mock.Mock(spec=aiohttp.ClientResponse)
        r1.status_code = 201
        r1.headers = {"location": "http://example/api/v1/test/1"}
        r1.content = ''

        r2 = mock.Mock(spec=aiohttp.ClientResponse)
        r2.status_code = 200
        r2.headers = {"content-type": "application/json"}
        r2.content = '{"result": ["a", "b", "c"]}'

        self.base_resource._store.update({
            "session": mock.Mock(spec=aiohttp.ClientSession),
            "serializer": slumber.serialize.Serializer(),
        })
        self.base_resource._store["session"].request.side_effect = (r1, r2)

        resp = self.base_resource._request("POST")

        self.assertTrue(resp is r1)
        self.assertEqual(resp.content, r1.content)

        self.base_resource._store["session"].request.assert_called_once_with(
            "POST",
            "http://example/api/v1/test",
            data=None,
            files=None,
            params=None,
            headers={"accept": self.base_resource._store["serializer"].get_content_type()}
        )

        resp = self.base_resource.post(data={'foo': 'bar'})
        self.assertEqual(resp['result'], ['a', 'b', 'c'])

    def test_post_decodable_response(self):
        r = mock.Mock(spec=aiohttp.ClientResponse)
        r.status_code = 200
        r.content = '{"result": ["a", "b", "c"]}'
        r.headers = {"content-type": "application/json"}

        self.base_resource._store.update({
            "session": mock.Mock(spec=aiohttp.ClientSession),
            "serializer": slumber.serialize.Serializer(),
        })
        self.base_resource._store["session"].request.return_value = r

        resp = self.base_resource._request("POST")

        self.assertTrue(resp is r)
        self.assertEqual(resp.content, r.content)

        self.base_resource._store["session"].request.assert_called_once_with(
            "POST",
            "http://example/api/v1/test",
            data=None,
            files=None,
            params=None,
            headers={"accept": self.base_resource._store["serializer"].get_content_type()}
        )

        resp = self.base_resource.post(data={'foo': 'bar'})
        self.assertEqual(resp['result'], ['a', 'b', 'c'])

    def test_patch_201_redirect(self):
        r1 = mock.Mock(spec=aiohttp.ClientResponse)
        r1.status_code = 201
        r1.headers = {"location": "http://example/api/v1/test/1"}
        r1.content = ''

        r2 = mock.Mock(spec=aiohttp.ClientResponse)
        r2.status_code = 200
        r2.headers = {"content-type": "application/json"}
        r2.content = '{"result": ["a", "b", "c"]}'

        self.base_resource._store.update({
            "session": mock.Mock(spec=aiohttp.ClientSession),
            "serializer": slumber.serialize.Serializer(),
        })
        self.base_resource._store["session"].request.side_effect = (r1, r2)

        resp = self.base_resource._request("PATCH")

        self.assertTrue(resp is r1)
        self.assertEqual(resp.content, r1.content)

        self.base_resource._store["session"].request.assert_called_once_with(
            "PATCH",
            "http://example/api/v1/test",
            data=None,
            files=None,
            params=None,
            headers={"accept": self.base_resource._store["serializer"].get_content_type()}
        )

        resp = self.base_resource.patch(data={'foo': 'bar'})
        self.assertEqual(resp['result'], ['a', 'b', 'c'])

    def test_patch_decodable_response(self):
        r = mock.Mock(spec=aiohttp.ClientResponse)
        r.status_code = 200
        r.content = '{"result": ["a", "b", "c"]}'
        r.headers = {"content-type": "application/json"}

        self.base_resource._store.update({
            "session": mock.Mock(spec=aiohttp.ClientSession),
            "serializer": slumber.serialize.Serializer(),
        })
        self.base_resource._store["session"].request.return_value = r

        resp = self.base_resource._request("PATCH")

        self.assertTrue(resp is r)
        self.assertEqual(resp.content, r.content)

        self.base_resource._store["session"].request.assert_called_once_with(
            "PATCH",
            "http://example/api/v1/test",
            data=None,
            files=None,
            params=None,
            headers={"accept": self.base_resource._store["serializer"].get_content_type()}
        )

        resp = self.base_resource.patch(data={'foo': 'bar'})
        self.assertEqual(resp['result'], ['a', 'b', 'c'])

    def test_put_201_redirect(self):
        r1 = mock.Mock(spec=aiohttp.ClientResponse)
        r1.status_code = 201
        r1.headers = {"location": "http://example/api/v1/test/1"}
        r1.content = ''

        r2 = mock.Mock(spec=aiohttp.ClientResponse)
        r2.status_code = 200
        r2.headers = {"content-type": "application/json"}
        r2.content = '{"result": ["a", "b", "c"]}'

        self.base_resource._store.update({
            "session": mock.Mock(spec=aiohttp.ClientSession),
            "serializer": slumber.serialize.Serializer(),
        })
        self.base_resource._store["session"].request.side_effect = (r1, r2)

        resp = self.base_resource._request("PUT")

        self.assertTrue(resp is r1)
        self.assertEqual(resp.content, r1.content)

        self.base_resource._store["session"].request.assert_called_once_with(
            "PUT",
            "http://example/api/v1/test",
            data=None,
            files=None,
            params=None,
            headers={"accept": self.base_resource._store["serializer"].get_content_type()}
        )

        resp = self.base_resource.put(data={'foo': 'bar'})
        self.assertEqual(resp['result'], ['a', 'b', 'c'])

    def test_put_decodable_response(self):
        r = mock.Mock(spec=aiohttp.ClientResponse)
        r.status_code = 200
        r.content = '{"result": ["a", "b", "c"]}'
        r.headers = {"content-type": "application/json"}

        self.base_resource._store.update({
            "session": mock.Mock(spec=aiohttp.ClientSession),
            "serializer": slumber.serialize.Serializer(),
        })
        self.base_resource._store["session"].request.return_value = r

        resp = self.base_resource._request("PUT")

        self.assertTrue(resp is r)
        self.assertEqual(resp.content, r.content)

        self.base_resource._store["session"].request.assert_called_once_with(
            "PUT",
            "http://example/api/v1/test",
            data=None,
            files=None,
            params=None,
            headers={"accept": self.base_resource._store["serializer"].get_content_type()}
        )

        resp = self.base_resource.put(data={'foo': 'bar'})
        self.assertEqual(resp['result'], ['a', 'b', 'c'])

    def test_handle_serialization(self):
        self.base_resource._store.update({
            "serializer": slumber.serialize.Serializer(),
        })

        resp = mock.Mock(spec=aiohttp.ClientResponse)
        resp.status_code = 200
        resp.headers = {"content-type": "application/json; charset=utf-8"}
        resp.content = '{"foo": "bar"}'

        r = self.base_resource._try_to_serialize_response(resp)

        if not isinstance(r, dict):
            self.fail("Serialization did not take place")

    def test_post_204_json(self):
        resp = mock.Mock(spec=aiohttp.ClientResponse)
        resp.status_code = 204
        resp.headers = {"content-type": "application/json"}
        resp.content = None

        self.base_resource._store.update({
            "session": mock.Mock(spec=aiohttp.ClientSession),
            "serializer": slumber.serialize.Serializer(),
        })

        self.base_resource._store["session"].request.return_value = resp

        self.assertEqual(self.base_resource.post(), None)

    def test_get_200_subresource_json(self):
        r = mock.Mock(spec=aiohttp.ClientResponse)
        r.status_code = 200
        r.headers = {"content-type": "application/json"}
        r.content = '{"result": ["a", "b", "c"]}'

        self.base_resource._store.update({
            "session": mock.Mock(spec=aiohttp.ClientSession),
            "serializer": slumber.serialize.Serializer(),
        })
        self.base_resource._store["session"].request.return_value = r

        resp = self.base_resource.subresource._request("GET")

        self.assertTrue(resp is r)
        self.assertEqual(resp.content, r.content)

        self.base_resource._store["session"].request.assert_called_once_with(
            "GET",
            "http://example/api/v1/test/subresource",
            data=None,
            files=None,
            params=None,
            headers={"accept": self.base_resource._store["serializer"].get_content_type()}
        )

        resp = self.base_resource.get()
        self.assertEqual(resp['result'], ['a', 'b', 'c'])

    def test_bad_resource_name(self):
        with self.assertRaises(AttributeError):
            self.base_resource._subresource

    def test_get_400_response(self):
        r = mock.Mock(spec=aiohttp.ClientResponse)
        r.status_code = 400
        r.headers = {"content-type": "application/json"}
        r.content = ''

        self.base_resource._store.update({
            "session": mock.Mock(spec=aiohttp.ClientSession),
            "serializer": slumber.serialize.Serializer(),
        })

        self.base_resource._store["session"].request.return_value = r

        with self.assertRaises(exceptions.HttpClientError):
            self.base_resource.req._request("GET")


    def test_get_404_response(self):
        r = mock.Mock(spec=aiohttp.ClientResponse)
        r.status_code = 404
        r.headers = {"content-type": "application/json"}
        r.content = ''

        self.base_resource._store.update({
            "session": mock.Mock(spec=aiohttp.ClientSession),
            "serializer": slumber.serialize.Serializer(),
        })
        self.base_resource._store["session"].request.return_value = r

        with self.assertRaises(exceptions.HttpNotFoundError):
            self.base_resource.req._request("GET")

    def test_get_500_response(self):
        r = mock.Mock(spec=aiohttp.ClientResponse)
        r.status_code = 500
        r.headers = {"content-type": "application/json"}
        r.content = ''

        self.base_resource._store.update({
            "session": mock.Mock(spec=aiohttp.ClientSession),
            "serializer": slumber.serialize.Serializer(),
        })
        self.base_resource._store["session"].request.return_value = r

        with self.assertRaises(exceptions.HttpServerError):
            self.base_resource.req._request("GET")

    def test_improperly_conf(self):
        with self.assertRaises(exceptions.ImproperlyConfigured):
            client = slumber.API()

    def test_api(self):
        r = mock.Mock(spec=aiohttp.ClientResponse)
        r.status_code = 200
        r.headers = {"content-type": "application/json"}
        r.content = '{"result": ["a", "b", "c"]}'

        client = slumber.API(base_url="http://example/api/v1", session=mock.Mock(spec=aiohttp.ClientSession))
        client.test._store["session"].request.return_value = r
        resp = client.test.get()

        self.assertEqual(resp['result'], ['a', 'b', 'c'])

    def test_api_subclass(self):
        class SubclassedResource(slumber.Resource):
            pass

        class SubclassedAPI(slumber.API):
            resource_class = SubclassedResource

        client = SubclassedAPI(base_url="http://example/api/v1")

        self.assertIsInstance(client.test, SubclassedResource)
        self.assertIsInstance(client.test(1).other(2).more, SubclassedResource)

    def test_url(self):
        self.assertEqual(self.base_resource.url(), "http://example/api/v1/test")

    def test_get_200_json_py3(self):
        r = mock.Mock(spec=aiohttp.ClientResponse)
        r.status_code = 200
        r.headers = {"content-type": "application/json"}
        r.content = b'{"result": ["a", "b", "c"]}'

        self.base_resource._store.update({
            "session": mock.Mock(spec=aiohttp.ClientSession),
            "serializer": slumber.serialize.Serializer(),
        })
        self.base_resource._store["session"].request.return_value = r

        resp = self.base_resource._request("GET")

        self.assertTrue(resp is r)
        self.assertEqual(resp.content, r.content)

        self.base_resource._store["session"].request.assert_called_once_with(
            "GET",
            "http://example/api/v1/test",
            data=None,
            files=None,
            params=None,
            headers={"accept": self.base_resource._store["serializer"].get_content_type()}
        )

        resp = self.base_resource.get()
        self.assertEqual(resp['result'], ['a', 'b', 'c'])

    def test_get_with_raw(self):
        r = mock.Mock(spec=aiohttp.ClientResponse)
        r.status_code = 200
        r.headers = {"content-type": "application/json"}
        r.content = '{"result": "a"}'

        self.base_resource._store.update({
            "session": mock.Mock(spec=aiohttp.ClientSession),
            "serializer": slumber.serialize.Serializer(),
            "raw": True,
        })
        self.base_resource._store["session"].request.return_value = r

        (response, decoded) = self.base_resource.get()

        self.assertIsInstance(response, aiohttp.ClientResponse)
        self.assertEqual(decoded["result"], "a")

    def test_as_raw_resource_get(self):
        apiurl = "http://example/api/v1"
        ses = mock.Mock(spec=aiohttp.ClientSession)
        r = mock.Mock(spec=aiohttp.ClientResponse)
        r.status_code = 200
        r.headers = {}
        ses.request.return_value = r

        api = slumber.API(apiurl, session=ses)

        (response, _) = api.myresource(1).subresource.as_raw().get()
        self.assertIsInstance(response, aiohttp.ClientResponse)

    def test_all_resource_requests_are_raw_if_set_in_api(self):
        apiurl = "http://example/api/v1"
        ses = mock.Mock(spec=aiohttp.ClientSession)
        r = mock.Mock(spec=aiohttp.ClientResponse)
        r.status_code = 200
        r.headers = {}
        ses.request.return_value = r

        api = slumber.API(apiurl, session=ses, raw=True)

        (response, _) = api.myresource(1).subresource.get()
        self.assertIsInstance(response, aiohttp.ClientResponse)

        (response, _) = api.myresource(1).get()
        self.assertIsInstance(response, aiohttp.ClientResponse)

    def test_send_content_type_only_if_body_data_exists(self):
        apiuri = "http://example/api/v1/"
        newuri = "http://example/api/v1/myresource/"
        ses = mock.Mock(spec=aiohttp.ClientSession)
        r = mock.Mock(spec=aiohttp.ClientResponse)
        r.status_code = 201
        r.headers = {}
        ses.request.return_value = r

        api = slumber.API(apiuri, session=ses)

        # Empty post request
        api.myresource.post()
        ses.return_value.status_code = 201
        ses.return_value.headers = {}
        self.assertEqual(ses.request.call_count, 1)

        ses.request.assert_called_with('POST', newuri,
                headers={
                    'accept': 'application/json'
                },
                data=None,
                files=None,
                params={})

        api.myresource.post(data=dict(key='value'))
        self.assertEqual(ses.request.call_count, 2)
        ses.request.assert_called_with('POST', newuri,
                headers={
                    'accept': 'application/json',
                    'content-type': 'application/json'
                },
                data='{"key": "value"}',
                files=None,
                params={})

    @unittest.expectedFailure
    def test_post_201_does_get(self):
        getparams = dict(username="luser", api_key="1234")
        postparams = dict(key1=1, key2="two")
        listuri = "http://example/api/v1/"
        newuri = "http://example/api/v1/myres/newthing/"
        ses = mock.Mock(spec=aiohttp.ClientSession)
        ses.request.return_value.status_code = 201
        ses.request.return_value.headers = { "location": newuri }
        api = slumber.API(listuri, session=ses)
        api.myres.post(postparams, **getparams)
        self.assertEqual(ses.request.call_count, 2)
        ses.request.assert_called_with('GET', newuri,
                headers={
                    'content-type': 'application/json',
                    'accept': 'application/json'
                    },
                params=getparams,
                data=None)

    def test_unicode_decodable_response(self):
        r = mock.Mock(spec=aiohttp.ClientResponse)
        r.status_code = 200
        r.content = '{"result": "Préparatoire"}'
        r.headers = {"content-type": "application/json"}

        self.base_resource._store.update({
            "session": mock.Mock(spec=aiohttp.ClientSession),
            "serializer": slumber.serialize.Serializer(),
        })
        self.base_resource._store["session"].request.return_value = r

        resp = self.base_resource._request("POST")

        self.assertTrue(resp is r)
        self.assertEqual(resp.content, r.content)

        self.base_resource._store["session"].request.assert_called_once_with(
            "POST",
            "http://example/api/v1/test",
            data=None,
            files=None,
            params=None,
            headers={"accept": self.base_resource._store["serializer"].get_content_type()}
        )

        resp = self.base_resource.post(data={'foo': 'bar'})
        expected = b'Pr\xc3\xa9paratoire'.decode('utf8')
        self.assertEqual(resp['result'], expected)


slumber/__init__.py
import aiohttp

try:
    from urllib.parse import urlparse, urlsplit, urlunsplit
except ImportError:
    from urlparse import urlparse, urlsplit, urlunsplit

from . import exceptions
from .serialize import Serializer
from .utils import url_join, iterator, copy_kwargs

__all__ = ["Resource", "API"]


class ResourceAttributesMixin(object):
    """
    A Mixin that allows access to an undefined attribute on a class.
    Instead of raising an attribute error, the undefined attribute will
    return a Resource Instance which can be used to make calls to the
    resource identified by the attribute.

    The type of the resource returned can be overridden by adding a
    resource_class attribute.

    It assumes that a Meta class exists at self._meta with all the required
    attributes.
    """

    def __getattr__(self, item):
        # Don't allow access to 'private' by convention attributes.
        # @@@: How would this work with resources names that begin with
        # underscores?
        if item.startswith("_"):
            raise AttributeError(item)

        kwargs = copy_kwargs(self._store)
        kwargs.update({"base_url": url_join(self._store["base_url"], item)})

        return self._get_resource(**kwargs)


class Resource(ResourceAttributesMixin, object):
    """
    Resource provides the main functionality behind slumber. It handles the
    attribute -> url, kwarg -> query param, and other related behind the scenes
    python to HTTP transformations. It's goal is to represent a single resource
    which may or may not have children.

    It assumes that a Meta class exists at self._meta with all the required
    attributes.
    """

    def __init__(self, *args, **kwargs):
        self._store = kwargs

    def __call__(self, id=None, format=None, url_override=None):
        """
        Returns a new instance of self modified by one or more of the available
        parameters. These allows us to do things like override format for a
        specific request, and enables the api.resource(ID).get() syntax to get
        a specific resource by it's ID.
        """

        # Short Circuit out if the call is empty
        if id is None and format is None and url_override is None:
            return self

        kwargs = copy_kwargs(self._store)

        if id is not None:
            kwargs["base_url"] = url_join(self._store["base_url"], id)

        if format is not None:
            kwargs["format"] = format

        if url_override is not None:
            # @@@ This is hacky and we should probably figure out a better way
            #    of handling the case when a POST/PUT doesn't return an object
            #    but a Location to an object that we need to GET.
            kwargs["base_url"] = url_override

        kwargs["session"] = self._store["session"]

        return self._get_resource(**kwargs)

    def _request(self, method, data=None, files=None, params=None):
        serializer = self._store["serializer"]
        url = self.url()

        headers = {"accept": serializer.get_content_type()}

        if not files:
            if data is not None:
                headers["content-type"] = serializer.get_content_type()
                data = serializer.dumps(data)

        resp = self._store["session"].request(method, url, data=data, params=params, files=files, headers=headers)

        if 400 <= resp.status_code <= 499:
            exception_class = exceptions.HttpNotFoundError if resp.status_code == 404 else exceptions.HttpClientError
            raise exception_class("Client Error %s: %s" % (resp.status_code, url), response=resp, content=resp.content)
        elif 500 <= resp.status_code <= 599:
            raise exceptions.HttpServerError("Server Error %s: %s" % (resp.status_code, url), response=resp, content=resp.content)

        self._ = resp

        return resp

    def _handle_redirect(self, resp, **kwargs):
        # @@@ Hacky, see description in __call__
        resource_obj = self(url_override=resp.headers["location"])
        return resource_obj.get(**kwargs)

    def _try_to_serialize_response(self, resp):
        s = self._store["serializer"]
        if resp.status_code in [204, 205]:
            return

        if resp.headers.get("content-type", None) and resp.content:
            content_type = resp.headers.get("content-type").split(";")[0].strip()

            try:
                stype = s.get_serializer(content_type=content_type)
            except exceptions.SerializerNotAvailable:
                return resp.content

            return stype.loads(resp.content)
        else:
            return resp.content

    def _process_response(self, resp):
        if 200 <= resp.status_code <= 299:
            decoded = self._try_to_serialize_response(resp)
        else:
            # @@@ We should probably do some sort of error here? (Is this even possible?)
            decoded = None

        if self._store["raw"]:
            return (resp, decoded)

        return decoded

    def _do_verb_request(self, verb, data=None, files=None, params=None):
        resp = self._request(verb, data=data, files=files, params=params)
        return self._process_response(resp)

    def as_raw(self):
        self._store["raw"] = True
        return self

    def get(self, **kwargs):
        return self._do_verb_request("GET", params=kwargs)

    def options(self, **kwargs):
        return self._do_verb_request("OPTIONS", params=kwargs)

    def head(self, **kwargs):
        return self._do_verb_request("HEAD", params=kwargs)

    def post(self, data=None, files=None, **kwargs):
        return self._do_verb_request("POST", data=data, files=files, params=kwargs)

    def patch(self, data=None, files=None, **kwargs):
        return self._do_verb_request("PATCH", data=data, files=files, params=kwargs)

    def put(self, data=None, files=None, **kwargs):
        return self._do_verb_request("PUT", data=data, files=files, params=kwargs)

    def delete(self, **kwargs):
        resp = self._request("DELETE", params=kwargs)
        if 200 <= resp.status_code <= 299:
            if resp.status_code == 204:
                return True
            else:
                return True  # @@@ Should this really be True?
        else:
            return False

    def url(self):
        url = self._store["base_url"]

        if self._store["append_slash"] and not url.endswith("/"):
            url = url + "/"

        return url


    def _get_resource(self, **kwargs):
        return self.__class__(**kwargs)


class API(ResourceAttributesMixin, object):

    resource_class = Resource

    def __init__(self, base_url=None, auth=None,
                 format=None, append_slash=True,
                 session=None, serializer=None, raw=False):
        if serializer is None:
            serializer = Serializer(default=format)

        if session is None:
            session = aiohttp.ClientSession()

        if auth is not None:
            session.auth = auth

        self._store = {
            "base_url": base_url,
            "format": format if format is not None else "json",
            "append_slash": append_slash,
            "session": session,
            "serializer": serializer,
            "raw": raw,
        }

        # Do some Checks for Required Values
        if self._store.get("base_url") is None:
            raise exceptions.ImproperlyConfigured("base_url is required")

    def _get_resource(self, **kwargs):
        return self.resource_class(**kwargs)
